File: Sorting/Pigeonhole_Sort/Pigeonhole_Sort.c
#include<stdio.h>
#include<stdlib.h>
 
void pigeonholesort(int arr[], int size) 
{ 
   int min = arr[0], max = arr[0]; 
   // Finding min value and max values in the array arr[]
   for (int i = 1; i < size; i++) 
   { 
        if (arr[i] < min) 
            min = arr[i]; 
        if (arr[i] > max) 
            max = arr[i]; 
   } 
   int range;
   range = max - min + 1;
   //Creating the holes
   int holes[range];
   for(int i=0;i<range;i++)
           holes[i]=0;
   //Filling the holes
   for (int j=0; j < size;j++)
          holes[arr[j] - min]++;
 
   // Put the elements back into the array in order.
   // k is the index of sorted array
   int k=0;
   for (int count = 0; count < range; count++)
      while (holes[count]-- > 0)
        {
              arr[k] = count + min;
              k=k+1;
        }
}
int main()
{ 
   int size,i;
   printf("Enter the size of the array ");
   scanf("%d",&size);
   int arr[size];
   printf("Enter the elements ");
   for ( i=0;i<size;i++)
       scanf("%d",&arr[i]) ;
   
   pigeonholesort(arr, size);
   
   printf("Sorted array :\n");
   for ( i = 0; i < size; i++)
         printf("%d ",arr[i]);
                        
   return 0;
}

//INPUT:Enter the size of the array 5
//Enter the elements -10
//5
//0
//9
//8
//OUTPUT:Sorted array: 
//-10 0 5 8 9 

File: Pigeonhole_Sort/Pigeonhole_Sort.js
//taking dynamic input array to be sorted
var i;
var n = prompt("enter the size of your array");
var arr = [];

for ( i = 0 ; i < n ; i++) 
{
    //Taking Input from user
    arr[i] = prompt('Enter Element ' + (i + 1));
}

//function for pigeonsort
function pigeonsort(array)
{
    //finding maximum value of array 
    var max = Math.max.apply(null,array); //19
  
    //finding minimum value of array 
    var min = Math.min.apply(null,array); //2
  
    //finding the size of hole array to be created
    var holerange = max - min + 1;
  
    //initialising hole array with value 0
    var hole = Array(holerange).fill(0);
  
    //filling the hole array
    for (i = 0 ; i < array.length ; i++)
    {
        hole[array[i] - min]++;
    }
	
    count = 0;
  
    //finally putting the values in sorted manner
    for (var j = 0 ; j < holerange ; j++)
    {
        while (hole[j] !== 0)
        {
            array[count] = j + min;
            hole[j]--;
      	    count++;
    	}
    }
    return array;
}

console.log(pigeonsort(arr));

//input -  [19,9,2,3,8,12,10]
//output - [2,3,8,9,10,12,19]
